sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoin.scala

```diff
@@ -84,84 +129,164 @@ case class SortMergeJoin(
-            rightElement = null
+            bufferedElement = null
           }
         }
 
         private def initialize() = {
-          fetchLeft()
-          fetchRight()
+          fetchStreamed()
+          fetchBuffered()
         }
 
         /**
          * Searches the right iterator for the next rows that have matches in left side, and store
          * them in a buffer.
+         * When this is not a Inner join, we will also return true when we get a row with no match
+         * on the other side. This search will jump out every time from the same position until
+         * `next()` is called.
          *
          * @return true if the search is successful, and false if the right iterator runs out of
          *         tuples.
          */
         private def nextMatchingPair(): Boolean = {
-          if (!stop && rightElement != null) {
-            // run both side to get the first match pair
-            while (!stop && leftElement != null && rightElement != null) {
-              val comparing = keyOrdering.compare(leftKey, rightKey)
+          if (!stop && streamedElement != null) {
+            // step 1: run both side to get the first match pair
+            while (!stop && streamedElement != null && bufferedElement != null) {
+              val comparing = keyOrdering.compare(streamedKey, bufferedKey)
               // for inner join, we need to filter those null keys
-              stop = comparing == 0 && !leftKey.anyNull
-              if (comparing > 0 || rightKey.anyNull) {
-                fetchRight()
-              } else if (comparing < 0 || leftKey.anyNull) {
-                fetchLeft()
+              stop = comparing == 0 && !streamedKey.anyNull
+              if (comparing > 0 || bufferedKey.anyNull) {
+                if (joinType == FullOuter) {
+                  // the join type is full outer and the buffered side has a row with no
+                  // join match, so we have a result row with streamed null with buffered
+                  // side as this row. Then we fetch next buffered element and go back.
+                  continueStreamed = false
+                  return true
+                } else {
+                  fetchBuffered()
+                }
+              } else if (comparing < 0 || streamedKey.anyNull) {
+                if (joinType == Inner) {
+                  fetchStreamed()
+                } else {
+                  // the join type is not inner and the streamed side has a row with no
+                  // join match, so we have a result row with this streamed row with buffered
+                  // null row. Then we fetch next streamed element and go back.
+                  continueStreamed = true
+                  return true
+                }
               }
             }
-            rightMatches = new CompactBuffer[Row]()
+            // step 2: run down the buffered side to put all matched rows in a buffer
+            bufferedMatches = new CompactBuffer[Row]()
```

Comment:
It seems like we end up allocating a lot of `CompactBuffers` over the lifetime of this merging task.  What do you think about adding a `clear()` method to `CompactBuffer` which nulls out the two fields and `Array.fill`s `null` into the portion of the array that was written to (if it's non-null)?  This could result in much less allocation / garbage here.